### ISSUE TITLE
user namespace setup support

### ISSUE THREAD

**Issue by trws** [flux-framework/flux-core#3927]
On LLNL TOSS3 systems  users can invoke a slurm plugin with `--userns` that configures the compute nodes they are assigned to allow them to use user/fs/etc namespaces.  The main use-case for this is running container software like podman and buildah, or sandboxed apps like flatpacks or snaps or similar.  As it stands now, TOSS4 has no similar option.

As I see it, there are two ways this could get resolved.  If user namespaces for regular users are enabled by default on TOSS4, then we probably don't need to do anything in flux at all.  That is currently not the case, but I'm not sure why it couldn't be.  If the plan is to keep it locked down like in TOSS3, then we'll need this for some of the container building workflows we're starting to stand up.

**Comment by dongahn** [comment]
Tagging @tpatki since this can be related to her FY22 ISCP project.

**Comment by grondo** [comment]
If this requires a call to `unshare()` then we may have to implement this an an IMP "plugin" - a facility that unfortunately does not yet exist. Do you know where the source of the `userns` plugin lives so we can see what other setup might be required?

I wonder if there is a strong argument for enabling user namespaces by default on the compute nodes, though, since every job can conceivably use `--userns` and get that support anyway? I'm not aware enough of the side-effects of using a user namespace to know if users would want to normally run without them enabled if they are not needed, though.

**Comment by trws** [comment]
As a user, there is no reason at all to ever have the support unavailable.  It costs nothing.  The only reason to keep it turned off is if someone thinks there's a potential security or other issue, but if we can just turn it on with slurm I'm not sure why that would be good enough.

**Comment by garlick** [comment]
Found this:  https://slurm.schedmd.com/SLUG19/NERSC_job_container.pdf

**Comment by garlick** [comment]
As noted in #4758, it seems that user namespaces are enabled on TOSS 4.  The question remains if that is sufficient to allow flux to launch full OS containers (like singularity) without adding any special support to the IMP or elsewhere.  Can we do that experiment and report what, if anything, is needed from Flux to support launching jobs in containers?  I think that is the question that motivated this issue.  The other cool namespacey things we can do if allowed can have their own issues as we dream them up.

**Comment by garlick** [comment]
@trws brought up idea of adding support to flux for mounting user-created file system images in private namespaces for jobs on demand.  He pointed out that [fuse2fs(1)](https://man7.org/linux/man-pages/man1/fuse2fs.1.html) (part of e2fsprogs) allows one to mount an ext4 block device image that was created by a user without the usual kernel oops/security concerns of doing that with a direct kernel ext4 mount.

He says the following runes were sufficient to make a fuse2fs mount possible.
```
unshare --user --mount --map-root-user <shell>
```

So maybe a shell plugin that just takes a path to an image file in a shared file system and mounts it in FLUX_JOB_TMPDIR?

**Comment by wihobbs** [comment]
> Can we do that experiment and report what, if anything, is needed from Flux to support launching jobs in containers? I think that is the question that motivated this issue. The other cool namespacey things we can do if allowed can have their own issues as we dream them up.
> > So maybe a shell plugin that just takes a path to an image file in a shared file system and mounts it in FLUX_JOB_TMPDIR?

Is there still a need to do this experiment and determine next steps for supporting user namespaces? Is that something we're willing or even want to support? Is anybody asking for it? I'm doing a pass of some of these issues to get an idea of where we're at.

Circling back after the fluxion meeting this week (when we discussed the work I'm doing with ephemeral ext4 files in lustre), It would be nice to have a root-less alternative, particularly if it also helps with container support or other areas we're exploring. But it seems like maybe we went down this road already, and decided against it?

**Comment by trws** [comment]
We still need a solution for ensuring users have additional UID mappings in the LC environment, ideally matching the slurm `--userns` option we have as a plugin on toss4 slurm systems.

**Comment by wihobbs** [comment]
@trws and I talked over email to Elena Green this morning, who wrote the `--userns` SPANK plugin for LC. Since LC has amended our subuid and subgid management strategy, we no longer have to support this to support containers in LC.

Ok to close this issue? We could also leave it open for some of the "nice to haves" that have been discussed here:

- container image pre-propagation to `/var/tmp` so that users aren't hammering the network filesystem fetching containers in parallel (a prolog might be able to do this quite trivially)
- a plugin that takes an ext4 device and uses `fuse2fs` to mount in a mount namespace

**Comment by garlick** [comment]
Nice!

For my education, how does one launch, say, an MPI hello world program in a container under Flux on an LC system?  Have we tried it?

**Comment by trws** [comment]
I have not yet, it sounds like they basically did what I was hoping for 
though which was to set up the sub*id mappings for everyone just by 
default on every node with a smaller range. If that’s right, then 
`flux run -N 5 -n 40 podman run some_registry:mpich_hello_world` should 
just go, or maybe after running the setup script.  It would also mean we 
could run _real_ containers on the login nodes for the first time, which 
would be fantastic.

**Comment by egreen77** [comment]
Yes, that's correct. However, there are a few issues with `podman` itself preventing MPI from working out-of-the-box like you describe. I've communicated those issues to the `podman` development team and hope to see progress soon.

**Comment by trws** [comment]
That’s good to know. If it isn’t sensitive, is it bootstrapping issues or NIC access or what kind of thing are we hitting?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by egreen77** [comment]
The main issue is that `podman` doesn't pass `fd`s from the job step launcher to the process inside the container, preventing the process from communicating with the PMI. 

Podman does have an option [`--preserve-fds=N`](https://docs.podman.io/en/latest/markdown/podman-run.1.html#preserve-fds-n) to pass in a block of `fd`s, but it expects those `fd`s to be in a contiguous range between 3 and 3+N and it'll abort if any of those `fd`s aren't present. The PMI communicator `fd`s that the runtime expects are usually in a disconnected range and don't play well with that. 

Newer versions of `podman` provide a `--preserve-fd` argument to preserve specific `fd`s, but that still requires some introspection at launch time by a wrapper script to determine the correct file descriptor values to provide to `podman`.

**Comment by garlick** [comment]
Ah, OK.   Flux uses "simple PMI" by default where the PMI_FD environment contains the single file descriptor number that's expected to be used by the MPI proc.

**Comment by trws** [comment]
That makes sense, we could probably help someone in writing a little OCI hook to handle that. May take a glance at it, I _think_ it would be pretty trivial since the FD number is accessible in the environment, but would have to check that the hook can access the option.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by trws** [comment]
Actually, digging around a bit this might be relevant: https://github.com/containers/podman/issues/10410#issuecomment-845125178
Mpich added a “—pmi-port” option to hydra to have it expose the PMI interface on a socket, if that uses the same simple protocol that might be really easy to add support for, maybe even expose both?  Have to see how the env var values change. Apparently there since MPICH 3.4.x at least.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by trws** [comment]
Sorry for the flood here, but it looks like it’s literally just set up a socket to accept connections, put the port number in PMI_PORT, and pass the resulting connected socket to the regular pmi-simple logic if someone connects. I kinda love PMI simple.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by garlick** [comment]
If that's the PMI_PORT stuff, I'm not sure it's going to be compiled-in on any of our MPI implementations.  It is not enabled by default, one has to configure mpich with `--enable-pmiport`.  It's also insecure (although we could bind to localhost and have it not be _too_ bad)

Hmm, the MPIs will  dlopen `libpmi.so` if PMI_FD is not set (might have to verify that).  That could internally use PMI_PORT if set.  An improvement over that might be to add a Flux RPC transport for the PMI server so that libpmi.so could securely connect to the PMI server via flux's `local://` connector.

PMIx already uses a unix domain socket so I would imagine  openmpi with pmix would just work here.  We're really only talking about the MPICH derivatives.

**Comment by trws** [comment]
Yeah, OpenMPI should work, but if we want the library to be loaded we would have to bind mount it into the container and do a great deal of extra intrusive work to do that. Even domain sockets would need to be bind mounted in, but at least wouldn’t require injecting a library that may or may not be compiled with a compatible libc.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by garlick** [comment]
Oh ick, good point.

**Comment by wihobbs** [comment]
@garlick to your earlier question, here is an example of how to run a super simple single-user container on a Flux-scheduled cluster in LC. Note that I did not try running an MPI hello world due to the limitations Tom brought up. But you can run python, and probably jupyterhub too (I didn't bother with VNC to actually see, although it said it launched successfully), which is something I've seen users clamoring for in mattermost.

```
flux alloc -N1
(s=1,d=1)  corona197 ~ $ /admin/scripts/weg/enable-podman.sh # do each time, adds things in /tmp
(s=1,d=1)  corona197 ~ $ podman pull jupyterhub/singleuser
Resolved "jupyterhub/singleuser" as an alias (/g/g0/hobbs17/.cache/containers/short-name-aliases.conf)
Trying to pull docker.io/jupyterhub/singleuser:latest...
Getting image source signatures
Copying blob 9bf717fe8843 done   |
...
Writing manifest to image destination
f00aa96928ba3b1a852cca75dc05b759e3893be67e225f29f231b6a730dc474b
WARN[0045] Failed to add pause process to systemd sandbox cgroup: dbus: couldn't determine address 
   of session bus
(s=1,d=1)  corona197 ~ $ podman run --rm -it jupyterhub/singleuser python3
Entered start.sh with args: python3
Running hooks in: /usr/local/bin/start-notebook.d as uid: 1000 gid: 100
Done running hooks in: /usr/local/bin/start-notebook.d
Running hooks in: /usr/local/bin/before-notebook.d as uid: 1000 gid: 100
Sourcing shell script: /usr/local/bin/before-notebook.d/10activate-conda-env.sh
Done running hooks in: /usr/local/bin/before-notebook.d
Executing the command: python3
Python 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>

```

I was unable to run OS containers like ubuntu due to an insufficient subgid space, which I guess is to be expected since every user only gets 2048:
```
root@7168706153ff:/bin# apt update
E: setgroups 65534 failed - setgroups (22: Invalid argument)
E: setegid 65534 failed - setegid (22: Invalid argument)
Reading package lists... Done
E: setgroups 65534 failed - setgroups (22: Invalid argument)
E: setegid 65534 failed - setegid (22: Invalid argument)
E: Method gave invalid 400 URI Failure message: Failed to setgroups - setgroups (22: Invalid argument)
E: Method gave invalid 400 URI Failure message: Failed to setgroups - setgroups (22: Invalid argument)
E: Method http has died unexpectedly!
E: Sub-process http returned an error code (112)
```

And I didn't try running MPI due to the limitations described above. But I suppose maybe something to try at a later date.

**Comment by wihobbs** [comment]
Sorry, the above note about ubuntu containers is...confusing (apologies, only 1 cup of coffee so far!) I could run them, but I'd say they're fairly useless, as they throw the error above whenever you try to install anything (like compilers for trying to build a hello-world).

**Comment by wihobbs** [comment]
Here's a little primer on running containers in LC: https://hpc.llnl.gov/services/cloud/containers. Some of the "gotchas" are probably different with the new way of managing sub[u/g]ids.

**Comment by egreen77** [comment]
We haven't updated the documentation to match the new subuid scheme yet. `apt` failed because Debian distros need either an `apt` config edit or a custom mapping because they use UID 65534 as a sandbox user. 

Try using:
```
$ podman run --uidmap 0:0:2000 --uidmap 65534:2047:1 ...
```

**Comment by trws** [comment]
I wonder if we could mitigate some of these with a couple of
hook scripts.  The MPI thing I'm going to try and play with today and see if I
can find a basic workaround.

**Comment by trws** [comment]
Ok, this is vaguely awful, but it does actually work on LC with mpich in podman today:

command: `flux run -N 2 -n 4 ./podman-test.sh`

script `podman-test.sh`:
```sh
 #!/bin/sh
 echo got $PMI_FD

 exec podman run -v "$(pwd):$(pwd)" \
   --net=host --pid=host --ipc=host \
   --preserve-fds="1" \
   -w "$(pwd)" --rm -it -u "0:0" \
   --env-host docker.io/mfisherman/mpich:latest \
   /bin/sh -c "PMI_FD=3 ./mpi-test" 3>&$PMI_FD
```

With a _slightly_ newer podman the FD manipulation isn't necessary, I tested this on my workstation and it's somewhat less painful:

```sh
 #!/bin/sh

 exec podman run -v "$(pwd):$(pwd)" \
   --net=host --pid=host --ipc=host \
   --preserve-fd="$PMI_FD" \
   -w "$(pwd)" --rm -it -u "0:0" \
   --env-host docker.io/mfisherman/mpich:latest \
   ./mpi-test
```

**Comment by trws** [comment]
FWIW I also hacked together a version of the script that uses socat to connect the FD to a port, and supplies the port along with the hostname to the executable.  Our mpich does in fact have PMI_PORT support, or at least it connects, but we get a pmi-simple wire protocol error so there must be something it asks for that we don't support because FD doesn't use it.

**Comment by garlick** [comment]
Oh that is good to know!

There is an additional 'fullinit' protocol element that PMI_PORT requires
since the peer rank is not known in advance like with PMI_FD.  I should
just go ahead and add that and document it.

**Comment by trws** [comment]
In case you want to play with it, here's the script:

```sh
#!/bin/bash
echo got $PMI_FD
env | grep PMI

export PORT=$((12345 + $PMI_RANK))
export PMI_PORT=localhost:$PORT
export PMI_ID=$PMI_RANK
socat -s -4 fd:$PMI_FD tcp-listen:$PORT &

PID=$!
unset PMI_FD
podman run -v "$(pwd):$(pwd)" \
  --net=host --pid=host --ipc=host \
  -w "$(pwd)" --rm -it -u "0:0" \
  --env-host docker.io/mfisherman/mpich:latest \
  /bin/sh -c "./mpi-test"
kill $PID
```

And this is the error with `-overbose=2`:
```
26.598s: flux-shell[0]: TRACE: pmi-simple: 2: C: cmd=initack pmiid=2
26.598s: flux-shell[0]: TRACE: pmi-simple: 2: S: pmi request error
26.598s: flux-shell[0]: FATAL: pmi-simple: PMI-1 wire protocol error
```
